File: 1-Part/4-chapter/mazeSolver.js
let MAZE = `
#######################################################################
#S#                 #       # #   #     #         #     #   #         #
# ##### ######### # ### ### # # # # ### # # ##### # ### # # ##### # ###
# #   #     #     #     #   # # #   # #   # #       # # # #     # #   #
# # # ##### # ########### ### # ##### ##### ######### # # ##### ### # #
#   #     # # #     #   #   #   #         #       #   #   #   #   # # #
######### # # # ##### # ### # ########### ####### # # ##### ##### ### #
#       # # # #     # #     # #   #   #   #     # # #   #         #   #
# # ##### # # ### # # ####### # # # # # # # ##### ### ### ######### # #
# # #   # # #   # # #     #     #   #   #   #   #   #     #         # #
### # # # # ### # # ##### ####### ########### # ### # ##### ##### ### #
#   # #   # #   # #     #   #     #       #   #     # #     #     #   #
# ### ####### ##### ### ### ####### ##### # ######### ### ### ##### ###
#   #         #     #     #       #   # #   # #     #   # #   # #   # #
### ########### # ####### ####### ### # ##### # # ##### # # ### # ### #
#   #   #       # #     #   #   #     #       # # #     # # #   # #   #
# ### # # ####### # ### ##### # ####### ### ### # # ####### # # # ### #
#     #         #     #       #           #     #           # #      E#
#######################################################################
`.split('\n')

const EMPTY = ' ';
const START = 'S';
const EXIT = 'E';
const PATH = '.';

const HEAIGHT = MAZE.length;
let maxWidthSoFar = MAZE[0].length;
for (let row of MAZE) {
    if (row.length > maxWidthSoFar) {
        maxWidthSoFar = row.length;
    }
}

const WIDTH = maxWidthSoFar
for (let i = 0; i < MAZE.length; i++) {
    MAZE[i] = MAZE[i].split('');
    if (MAZE[i].length !== WIDTH) {
        MAZE[i] = EMPTY.repeat(WIDTH).split('')
    }
}

function printMaze(maze) {
    // console.log('.........')
    // for (let y = 0; y < HEAIGHT; y++) {
    //     for (let x = 0; x < WIDTH; x++) {
    //         console.log(maze[y][x])
    //     }
    // }
    console.log(maze.map(row => row.join('')).join('\n'));
}

function findStart(maze) {
    for (let x = 0; x < WIDTH; x++) {
        for (let y = 0; y < HEAIGHT; y++) {
            if (maze[y][x] === START) {
                return [x, y]
            }
        }
    }
}

function solveMaze(maze, x, y, visited) {
    if (x === undefined || y === undefined) {
        [x, y] = findStart(maze)
        maze[y][x] = EMPTY
    }
    if (visited === undefined) {
        visited = []
    }

    if (maze[y][x] === EXIT) {
        return true
    }

    maze[y][x] = PATH
    visited.push(String(x) + ',' + String(y))

    // printMaze(maze)

    if ((y + 1 < HEAIGHT) && ((maze[y + 1][x] === EMPTY) || (maze[y + 1][x] === EXIT)) && (visited.indexOf(String(x) + ',' + String(y + 1)) === -1)) {
        if (solveMaze(maze, x, y + 1, visited)) {
            return true
        }
    }
    if ((y - 1 >= 0) && ((maze[y - 1][x] === EMPTY) || (maze[y - 1][x] === EXIT)) && (visited.indexOf(String(x) + ',' + String(y - 1)) === -1)) {
        if (solveMaze(maze, x, y - 1, visited)) {
            return true
        }
    }
    if ((x + 1 < WIDTH) && ((maze[y][x + 1] === EMPTY) || (maze[y][x + 1] === EXIT)) && (visited.indexOf(String(x + 1) + ',' + String(y)) === -1)) {
        if (solveMaze(maze, x + 1, y, visited)) {
            return true
        }
    }
    if ((x - 1 >= 0) && ((maze[y][x - 1] === EMPTY) || (maze[y][x - 1] === EXIT)) && (visited.indexOf(String(x - 1) + ',' + String(y)) === -1)) {
        if (solveMaze(maze, x - 1, y, visited)) {
            return true
        }
    }
    maze[y][x] = EMPTY
    // printMaze(maze)
    return false
}

printMaze(MAZE)
solveMaze(MAZE)
printMaze(MAZE)

File: 1-Part/4-chapter/mazeSolver.py
MAZE = """ 
#######################################################################
#S#                 #       # #   #     #         #     #   #         #
# ##### ######### # ### ### # # # # ### # # ##### # ### # # ##### # ###
# #   #     #     #     #   # # #   # #   # #       # # # #     # #   #
# # # ##### # ########### ### # ##### ##### ######### # # ##### ### # #
#   #     # # #     #   #   #   #         #       #   #   #   #   # # #
######### # # # ##### # ### # ########### ####### # # ##### ##### ### #
#       # # # #     # #     # #   #   #   #     # # #   #         #   #
# # ##### # # ### # # ####### # # # # # # # ##### ### ### ######### # #
# # #   # # #   # # #     #     #   #   #   #   #   #     #         # #
### # # # # ### # # ##### ####### ########### # ### # ##### ##### ### #
#   # #   # #   # #     #   #     #       #   #     # #     #     #   #
# ### ####### ##### ### ### ####### ##### # ######### ### ### ##### ###
#   #         #     #     #       #   # #   # #     #   # #   # #   # #
### ########### # ####### ####### ### # ##### # # ##### # # ### # ### #
#   #   #       # #     #   #   #     #       # # #     # # #   # #   #
# ### # # ####### # ### ##### # ####### ### ### # # ####### # # # ### #
#     #         #     #       #           #     #           # #      E#
#######################################################################
""".split('\n')

EMPTY = ' '
START = 'S'
EXIT = 'E'
PATH = '.'

HEIGHT = len(MAZE)
WIDTH = 0

for row in MAZE:
    if len(row) > WIDTH:
        WIDTH = len(row)
for i in range(len(MAZE)):
    MAZE[i] = list(MAZE[i])
    if len(MAZE[i]) is not WIDTH:
        MAZE[i] = [EMPTY] * WIDTH


def printMaze(maze):
    for y in range(HEIGHT):
        for x in range(WIDTH):
            print(maze[y][x], end='')
        print()
    print()


def findStart(maze):
    for x in range(WIDTH):
        for y in range(HEIGHT):
            if maze[y][x] == START:
                return (x, y)


def solveMaze(maze, x=None, y=None, vitited=None):
    if x == None and y == None:
        x, y = findStart(maze)
        maze[y][x] = EMPTY
    if vitited == None:
        vitited = []

    if maze[y][x] == EXIT:
        return True

    maze[y][x] = PATH
    vitited.append(str(x) + ',' + str(y))
    # printMaze(maze)

    if y + 1 < HEIGHT and maze[y + 1][x] in (EMPTY, EXIT) and str(x) + ',' + str(y + 1) not in vitited:
        if solveMaze(maze, x, y + 1, vitited):
            return True
    if y - 1 >= 0 and maze[y - 1][x] in (EMPTY, EXIT) and str(x) + ',' + str(y - 1) not in vitited:
        if solveMaze(maze, x, y - 1, vitited):
            return True
    if x + 1 < WIDTH and maze[y][x + 1] in (EMPTY, EXIT) and str(x + 1) + ',' + str(y) not in vitited:
        if solveMaze(maze, x + 1, y, vitited):
            return True
    if x - 1 >= 0 and maze[y][x - 1] in (EMPTY, EXIT) and str(x - 1) + ',' + str(y) not in vitited:
        if solveMaze(maze, x - 1, y, vitited):
            return True

    maze[y][x] = EMPTY
    # printMaze(maze)

    return False


printMaze(MAZE)
solveMaze(MAZE)
printMaze(MAZE)
